website/blog/2019-08-19-tablescan-structs.md

```diff
@@ -0,0 +1,163 @@
+---
+title: Table Scan: Doing The Right Thing With Structured Types
+author: Orri Erling
+authorURL: https://www.linkedin.com/in/orrierling/
+authorFBID: 100026224749124
+---
+
+In the previous article we saw what gains are possible when filtering early and in the right order. In this article we look at how we do this with nested and structured types.
+
+<!--truncate-->
+
+We use the 100G TPC-H dataset but now we group top level columns into structs or maps. 
+
+The tables are defined as follows:
+
+```sql
+CREATE TABLE hive.tpch.exportlineitem  (
+    l_orderkey BIGINT,
+    l_linenumber INTEGER,
+    l_shipment row(
+        l_partkey BIGINT,
+        l_suppkey BIGINT,
+        l_extendedprice DOUBLE,
+        l_discount DOUBLE,
+        l_quantity DOUBLE,
+        l_shipdate DATE,
+        l_receiptdate DATE,
+        l_commitdate DATE,
+        l_comment VARCHAR(44)),
+    l_export row(
+        s_nation BIGINT,
+        c_nation BIGINT,
+        is_inside_eu INTEGER,
+        is_restricted INTEGER,
+        license row(
+            applydate DATE,
+            grantdate DATE,
+            filing_no BIGINT,
+            comment VARCHAR))
+)
+WITH (
+    format = 'ORC'
+);
+```
+
+The l_shipment struct has the non-null top-level columns we all have come to know and love. The l_export struct is null if the customer and supplier nations are the same and present otherwise. A fraction of the rows have an additional nested export license struct.
+
+```sql
+CREATE TABLE hive.tpch.lineitem_map (
+    l_orderkey BIGINT,
+    l_linenumber INTEGER,
+    ints map(INTEGER, BIGINT),
+    strs map(VARCHAR(2), varchar(44))
+)
+WITH (
+    format = 'ORC'
+);
+```
+
+This table has a map of 12 integers in `ints` and 5 strings in `strs`. The key is the column ordinal number in the original lineitem table as an integer in `ints` and as a string in `strs`.
+
+Like before, the tables are in ORC V2 and are compressed with Snappy. We show the Aria and baseline times as wall time seconds / CPU seconds, labeled with A: and B: respectively. 
+
+```sql
+SELECT COUNT(*), SUM(l_shipment.l_extendedprice)
+FROM exportlineitem
+WHERE l_shipment.l_partkey BETWEEN 1000000 AND 2000000 AND l_shipment.l_suppkey BETWEEN 100000 AND 200000;
+```
+A: 4 / 34
+B:  18 / 253
+
+This experiment demonstrates that a non-null struct introduces no performance penalty as compared to using top level columns. The filters are pushed down into the column scans and no struct column is materialized.
```

Comment:
The numbers above do not show that. The reader would need to go back to previous blog and look up a matching query and numbers to confirm this statement. Would it be possible to include all the necessary info here?

Reply:
As part of the rework the query was brought into this blog post. Hopefully it is clearer now.